### PR TITLE
NAS-129929 / 24.10 / Improve SID handling

### PR DESCRIPTION
**Pull request by anodos325**
Do not rely on Samba to generate a new system SID and make randomized
SID persistent across server changes. This is to help prevent admin
foot-shooting when they choose to make major server changes that force
Samba to regenrate a new SID and thus invalidate their share ACLs.
on production servers.

Since local user / group RID values are deterministic based on the id
key for user / group accounts, populate `sid` key in user.query and
group.query extend methods. Apply similar logic to short-circuit
SID conversion.

The nt_name key provides little value to API consumers and so remove
for account entries.

Remove subprocess call to `net groupmap` in favor of using our tdb
utilities to directly alter the group_mapping.tdb file. This generally
performs better and avoids having to synchronize group mappings during
group CRUD methods.

### CODE REVIEW

File: src/middlewared/middlewared/plugins/idmap.py [hunk 1]
```diff
@@ -30,7 +36,6 @@
 WINBIND_IDMAP_FILE = '/var/run/samba-lock/gencache.tdb'
 WINBIND_IDMAP_TDB_OPTIONS = TDBOptions(TDBPathType.CUSTOM, TDBDataType.BYTES)
 
-
```
Comment (mgrimesix) [inline src/middlewared/middlewared/plugins/idmap.py:33]: Flake8 wants two lines here

File: src/middlewared/middlewared/plugins/idmap.py [hunk 2]
```diff
@@ -1067,6 +1075,59 @@ def __unixsid_to_name(self, sid, separator='\\'):
...
+        entry = self.middleware.call_sync(method, db_id)
+
+        return {
+            'name': f'{netbiosname}{separator}{entry[name_key]}',
+            'id': entry[xid_key],
+            'id_type': id_type,
+            'sid': sid
+        }
+
+
```
Comment (mgrimesix) [inline src/middlewared/middlewared/plugins/idmap.py:1130]: Flake8 wants one line (not two) here

File: src/middlewared/middlewared/plugins/smb_/groupmap.py [hunk 3]
```diff
@@ -332,168 +266,121 @@ async def groupmap_list(self):
...
-        payload = json.dumps(data)
-        out = await run([SMBCmd.NET.value, '--json', 'groupmap', 'batch', payload], check=False)
-        if out.returncode != 0:
-            raise CallError(f'Batch operation for [{data}] failed with error {out.stderr.decode()}')
+        return self.validate_groupmap_hwm(low_range)
 
     @private
-    @job(lock="groupmap_sync")
-    async def synchronize_group_mappings(self, job, bypass_sentinel_check=False):
+    @job(lock="groupmap_sync",  lock_queue_size=1)
```
Comment (mgrimesix) [inline src/middlewared/middlewared/plugins/smb_/groupmap.py:329]: Flake8 doesn't like two spaces: `,  lock`

File: src/middlewared/middlewared/pytest/unit/utils/test_groupmap.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+@pytest.fixture(scope='module')
+def groupmap_dir():
+    os.makedirs('/var/db/system/samba4', exist_ok=True)
+
+
+@pytest.fixture(scope='module')
+def local_sid():
+    try:
+       yield random_sid()
```
Comment (mgrimesix) [inline src/middlewared/middlewared/pytest/unit/utils/test_groupmap.py:30]: Indendation.   Here and lines 32 and 33.

File: tests/api2/test_smb_groupmap.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+        # Validate that re-enabling restores SID value
+        call('group.update', g['id'], {'smb': True})
+        new = call('group.get_instance', g['id'])
+        assert new['sid'] == g['sid']
+
+        groupmap = call('smb.groupmap_list')
+        assert str(new['gid']) in groupmap['local']
+
+@pytest.mark.parametrize('name,gid,sid', [
```
Comment (mgrimesix) [inline tests/api2/test_smb_groupmap.py:65]: Flake8 wants two lines

**Review by mgrimesix — COMMENTED** [review]
First pass:  Flake8 complaints.
There are unused imports in several modules, but not new.

File: src/middlewared/middlewared/pytest/unit/utils/test_groupmap.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    assert larry['sid_type'] == lsa_sidtype.ALIAS
+    assert larry['name'] == 'larry'
+    assert larry['comment'] == ''
+
+    delete_groupmap_entry(
+        GroupmapFile.DEFAULT,
+        GroupmapEntryType.GROUP_MAPPING,
+        f'{local_sid}-2000010'
+    )
+
```
Comment (bmeagherix) [inline src/middlewared/middlewared/pytest/unit/utils/test_groupmap.py:81]: Maybe make sure we can still get _larry_ now.

File: src/middlewared/middlewared/pytest/unit/utils/test_groupmap.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    assert res['sid'] == 'S-1-5-32-545'
+    assert set(res['members']) == set((f'{local_sid}-2000012', f'{local_sid}-2000013'))
+
+    delete_groupmap_entry(
+        GroupmapFile.DEFAULT,
+        GroupmapEntryType.MEMBERSHIP,
+        'S-1-5-32-544'
+    )
+
```
Comment (bmeagherix) [inline src/middlewared/middlewared/pytest/unit/utils/test_groupmap.py:139]: Again maybe ensure we can still get `-545` now.

File: src/middlewared/middlewared/pytest/unit/utils/test_sid.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+
+@pytest.mark.parametrize('sid,valid', [
+    ('S-1-5-21-3510196835-1033636670-2319939847-200108', True),
+    ('S-1-5-32-544', True),
+    ('S-1-2-0', False),  # technically valid SID but we don't permit it
+    ('S-1-5-21-3510196835-1033636670-2319939847-200108-200108', False),
+    ('S-1-5-21-3510196835-200108', False),
+    ('S-1-5-21-3510196835-1033636670-231993009847-200108', False),
+    ('S-1-5-21-351019683b-1033636670-231993009847-200108', False),
```
Comment (bmeagherix) [inline src/middlewared/middlewared/pytest/unit/utils/test_sid.py:42]: Not sure I can spot the difference between these two.  If there isn't one, did you mean to test something else?

File: tests/api2/test_smb_groupmap.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    members_dom_users = call('smb.groupmap_listmem', 'S-1-5-32-545')
+    assert entry['sid'] in members_dom_users
+
+
+def test__new_group():
+    with group({"name": "group1"}) as g:
+        # Validate GID is being assigned as expected
+        assert g['sid'] is not None
+        rid = int(g['sid'].split('-')[-1])
+        assert rid == g['id'] + BASE_RID_GROUP
```
Comment (bmeagherix) [inline tests/api2/test_smb_groupmap.py:41]: I don't _think_ I saw a **user** test similar to the above (where the `sid` & `rid` are checked).  Maybe add one?

**Review by bmeagherix — APPROVED** [review]
I'll approve as most of my comments are minor.

File: src/middlewared/middlewared/plugins/account.py [hunk 10]
```diff
@@ -268,15 +274,10 @@ def user_compress(self, user):
...
+        Query users with `query-filters` and `query-options`.
 
-        Expanded information may be requested by specifying the extra option
-        `"extra": {"additional_information": []}`.
-
-        The following `additional_information` options are supported:
-        `SMB` - include Windows SID and NT Name for user. If this option is not specified, then these
-            keys will have `null` value.
+        If users provided by Active Directory or LDAP are not desired, then
+        a "local", "=", False should be added to filters.
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/account.py:280]: Yes. I don't know why I keep inverting that :(

File: tests/api2/test_smb_groupmap.py (hunk 9 above)
Comment (anodos325) [reply tests/api2/test_smb_groupmap.py:41]: Yeah, that will be next major PR to improve passdb handling.

File: src/middlewared/middlewared/plugins/smb_/groupmap.py [hunk 11]
```diff
@@ -332,168 +266,121 @@ async def groupmap_list(self):
...
+
+        data = list_foreign_group_memberships(GroupmapFile.DEFAULT, sid)
+        assert data.sid == sid
+
+        return data.members
+
+    @private
+    def sync_builtins(self, to_add):
+        idmap_backend = self.middleware.call_sync("smb.getparm", "idmap config * : backend", "GLOBAL")
+        idmap_range = self.middleware.call_sync("smb.getparm", "idmap config * : range", "GLOBAL")
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/smb_/groupmap.py:306]: Nope. Shouldn't be possible, samba requires this to actually start.

**Pull request merged**